### ISSUE TITLE
Transfer Learning

### ISSUE THREAD

**Issue by iLOSTA** [iLOSTA/The-Project#9]
I have been reading about transfer learning and trying to figure out how to apply it to our project. As you suggested in the last meeting, I took 2 models 'prius and a5' out of the original dataset and trained the NN model. the model now makes predictions for 38 labels instead of 40. 

The method I am trying right now is to save the model to disk and load it again and try to train it on the new data (samples of toyota prius and audi a5), in which case the model needs to output 40 targets instead of 38. I can't seem to get it to work just yet because of the change in the output dimension.

**Comment by iLOSTA** [comment]
UPDATE 1:

I fixed the the input dimension difference problem by having a fixed input dimension for both old and new models. Logically when training the model the first time the dataset will be bigger than that when we want to expand the model later for new CAR MODELS for example. Thus, The the first classifier should not capture too many features to make sure it can take in the new data later.

**Comment by iLOSTA** [comment]
UPDATE 2:
From now on, I will be referring to two different datasets:
1- DATASET1 : Contains samples for 38 labels (Toyota Prius and Audi A5 taken out).
2- DATASET2 : Contains samples for 40 labels (Includes Prius and A5).

- The model was trained using DATASET1 and the model was saved on disk.
- Loaded the model and modified the output layer to make it output 40 labels and retrained the model using DATASET2.
- All the model's layers were retrained again (not frozen).
- The results are great. Now the model works for the new two CAR MODELS (Prius and A5).
Next I will try to freeze the old layers and only train the new output layer and see what happens.

**Comment by iLOSTA** [comment]
Freezing all the layers but the last one doesn't affect the process in any way. Same accuracy and same training time. That is probably because I am training using the entire dataset. I will try training with the new data + few samples of the the old data.

**Comment by iLOSTA** [comment]
UPDATE 3:
I created a 3rd Dataset that contains the new data plus 20% of the old data and used it to retrain the NN model. The result is not good at all. Although the validation accuracy appears to be high during training, there are a lot of miss-classified subcategories. 

My opinion on this:
If the goal of using transfer learning (in my project) is to reduce training time and save resources, then transfer learning is not of great use for this project. The reason is that in order to reduce the training time I have to either use less data or reduce training iterations (epochs). I tested both cases and the NN model performed poorly.
However, if the goal is to re-adapt the old NN model to make it work with new coming car models, car makes, new panels, etc., then we can use transfer learning but the new data plus ALL old data will have to be used for retraining.
